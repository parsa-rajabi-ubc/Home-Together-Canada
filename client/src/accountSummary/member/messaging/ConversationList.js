/**
 *
 * @Author:     Jeff Hatton
 * @Created:    2021.3.04
 *
 * @Description: Member Messaging Conversation Card List
 *
 */
import React, {useState}from 'react';
import PropTypes from "prop-types";
import Paginate from "../../../common/forms/Paginate";
import ConversationCard from "./ConversationCard";
import {mostRecentMessages,sortMessageByTimeDecreasing} from "./messageUtils";
import Confirmation from "../../../common/listings/Confirmation";
import EmptySelection from "./EmptySelection";
import FullConversation from "./FullConversation";

const NUM_RESULTS = 7;
const MESSAGE = {
    NO_RESULTS: "No results found"
}

function ConversationList(props) {
    const {messageData, messageUser} = props;
    const recentMessages = mostRecentMessages(messageData);
    sortMessageByTimeDecreasing(recentMessages);
    const cardData = [];
    const [isSelect,setIsSelect] = useState(false);
    const [senderId,setSenderId] = useState("");

    for (let i = 0; i < recentMessages.length; i++) {
        if (recentMessages[i].senderId===messageUser) {
            cardData.push(
                    <ConversationCard key={i}
                                      onClick={()=>{setIsSelect(true);setSenderId(recentMessages[i].receiverId)}}
                                      userName={recentMessages[i].receiverId}
                                      messageContent={recentMessages[i].messageContent}
                                      datePosted={recentMessages[i].dateSent}
                    />
            );
        }
        else if (recentMessages[i].receiverId===messageUser) {
            cardData.push(
                    <ConversationCard key={i}
                                      onClick={()=>{setIsSelect(true);setSenderId(recentMessages[i].senderId)}}
                                      userName={recentMessages[i].senderId}
                                      messageContent={recentMessages[i].messageContent}
                                      datePosted={recentMessages[i].dateSent}
                    />
            );
        }
    }

    return (
        <div>
            {(!cardData.length) ? <Confirmation displayButton={false} errorColor={true} message={MESSAGE.NO_RESULTS}/>
                : <Paginate data={cardData} resultsPerPage={NUM_RESULTS}/>}
            {(!isSelect || senderId ==="") ? <EmptySelection/>:<FullConversation messageData={messageData} myUserName={messageUser} senderId={senderId}/>}
        </div>
    );
}

ConversationList.propTypes = {
    messageUser: PropTypes.string.isRequired,
    messageData: PropTypes.array.isRequired,
};

export default ConversationList;