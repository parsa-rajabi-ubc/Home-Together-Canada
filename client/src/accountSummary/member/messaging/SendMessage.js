/**
 * @Author:     Jeff Hatton
 * @Created:    2021.03.10
 *
 * @Description: Member message sending component for profile page button
 *
 */

import React from 'react';
import PropTypes from "prop-types";
import LargeTextArea from "../../../common/forms/LargeTextArea";
import SubmitButton from "../../../common/forms/SubmitButton";

function SendMessage(props) {
    //username currently unused, but intended as component which will handle post request to add message to database
    const {username, otherUser} = props;
    function onSubmit(){
        return;
    }
    return(
        <div>
            <p>To {otherUser}: </p>
            <LargeTextArea label={"What would you like to contact this member about? "} />
            <SubmitButton className={"btn btn-green mb-6 w-1/2 text-base py-2"} onClick={onSubmit}/>
        </div>
    )
}

SendMessage.propTypes = {
    username: PropTypes.string.isRequired,
    otherUser: PropTypes.string.isRequired
};

export default SendMessage;