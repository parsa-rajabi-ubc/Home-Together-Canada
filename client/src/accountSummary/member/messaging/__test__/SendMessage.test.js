/**
 * @Author:     Jeff Hatton
 * @Created:    2021.3.12
 *
 * @Description: Conversation List Snapshot Test
 *
 */
import React from 'react';
import renderer from 'react-test-renderer'
import ConversationList from "../ConversationList";
import SendMessage from "../SendMessage";

jest.mock("react-tooltip/node_modules/uuid", () => ({
            v4: () => "00000000-0000-0000-0000-000000000000"
        }
    )
);
describe('ConversationList', () => {
    describe('Snapshot test', () => {
        it('should match snapshot test if listing is NOT empty for userName', () => {
            //given
            const props = {
                username: "messageMember1",
                otherUser: "otherUser1",
            }
            //when
            const component = renderer.create(<SendMessage {...props}/>);
            const tree = component.toJSON();
            //then
            expect(tree).toMatchSnapshot();
        });
    });
});