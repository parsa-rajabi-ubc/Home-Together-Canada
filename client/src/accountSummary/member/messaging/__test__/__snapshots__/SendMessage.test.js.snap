// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendMessage Snapshot test should match snapshot test of SendMessage 1`] = `
<div>
  <p>
    To 
    member2
    : 
  </p>
  <div>
    <label
      className="label"
    >
      What would you like to contact this member about? 
    </label>
    
    <textarea
      className="undefined undefined undefined input"
      cols="50"
      onChange={[Function]}
    />
  </div>
  <label>
    <input
      className="btn btn-green mb-6 w-1/2 text-base py-2"
      onClick={[Function]}
      type="submit"
    />
  </label>
</div>
`;
